File: Algorithms/BinarySearchTree.java

/**
 * Write a description of class i here.
 *
 * @author (your name)
 * @version (a version number or a date)
 */
// Java program to demonstrate delete operation in binary search tree 
class BinarySearchTree 
{ 
	/* Class containing left and right child of current node and key value*/
	class Node 
	{ 
		int key; 
		Node left, right; 

		public Node(int item) 
		{ 
			key = item; 
			left = right = null; 
		} 
	} 

	// Root of BST 
	Node root; 

	// Constructor 
	BinarySearchTree() 
	{ 
		root = null; 
	} 

	// This method mainly calls deleteRec() 
	void deleteKey(int key) 
	{ 
		root = deleteRec(root, key); 
	} 

	/* A recursive function to insert a new key in BST */
	Node deleteRec(Node root, int key) 
	{ 
		/* Base Case: If the tree is empty */
		if (root == null) return root; 

		/* Otherwise, recur down the tree */
		if (key < root.key) 
			root.left = deleteRec(root.left, key); 
		else if (key > root.key) 
			root.right = deleteRec(root.right, key); 

		// if key is same as root's key, then This is the node 
		// to be deleted 
		else
		{ 
			// node with only one child or no child 
			if (root.left == null) 
				return root.right; 
			else if (root.right == null) 
				return root.left; 

			// node with two children: Get the inorder successor (smallest 
			// in the right subtree) 
			root.key = minValue(root.right); 

			// Delete the inorder successor 
			root.right = deleteRec(root.right, root.key); 
		} 

		return root; 
	} 

	int minValue(Node root) 
	{ 
		int minv = root.key; 
		while (root.left != null) 
		{ 
			minv = root.left.key; 
			root = root.left; 
		} 
		return minv; 
	} 

	// This method mainly calls insertRec() 
	void insert(int key) 
	{ 
		root = insertRec(root, key); 
	} 

	/* A recursive function to insert a new key in BST */
	Node insertRec(Node root, int key) 
	{ 

		/* If the tree is empty, return a new node */
		if (root == null) 
		{ 
			root = new Node(key); 
			return root; 
		} 

		/* Otherwise, recur down the tree */
		if (key < root.key) 
			root.left = insertRec(root.left, key); 
		else if (key > root.key) 
			root.right = insertRec(root.right, key); 

		/* return the (unchanged) node pointer */
		return root; 
	} 

	// This method mainly calls InorderRec() 
	void inorder() 
	{ 
		inorderRec(root); 
	} 

	// A utility function to do inorder traversal of BST 
	void inorderRec(Node root) 
	{ 
		if (root != null) 
		{ 
			inorderRec(root.left); 
			System.out.print(root.key + " "); 
			inorderRec(root.right); 
		} 
	} 

	// Driver Program to test above functions 
	public static void main(String[] args) 
	{ 
		BinarySearchTree tree = new BinarySearchTree(); 

		/* Let us create following BST 
			50 
		/	 \ 
		30	 70 
		/ \ / \ 
		20 40 60 80 */
		tree.insert(50); 
		tree.insert(30); 
		tree.insert(20); 
		tree.insert(40); 
		tree.insert(70); 
		tree.insert(60); 
		tree.insert(80); 

		System.out.println("Inorder traversal of the given tree"); 
		tree.inorder(); 

		System.out.println("\nDelete 20"); 
		tree.deleteKey(20); 
		System.out.println("Inorder traversal of the modified tree"); 
		tree.inorder(); 

		System.out.println("\nDelete 30"); 
		tree.deleteKey(30); 
		System.out.println("Inorder traversal of the modified tree"); 
		tree.inorder(); 

		System.out.println("\nDelete 50"); 
		tree.deleteKey(50); 
		System.out.println("Inorder traversal of the modified tree"); 
		tree.inorder(); 
	} 
} 

File: Algorithms/BST/BinarySearchTree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.right = None
        self.left = None
        self.parent = None


class BinarySearchTree:
    def __init__(self):
        self.root = None

    # def search(t,x):
    #     if self.root !=None:
    #         if self.root.data == x:
    #             return self.root
    #         elif self.root.data > x:
    #             return self.search(self.root.left,x)
    #         else:
    #             return self.search(self.root.right,x)

    def minimum(self, x):
        while x.left != None:
            x = x.left
        return x

    def insert(self, n):
        y = None
        temp = self.root
        while temp != None:
            y = temp
            if n.data < temp.data:
                temp = temp.left
            else:
                temp = temp.right

        n.parent = y

        if y == None:  # newly added node is root
            self.root = n
        elif n.data < y.data:
            y.left = n
        else:
            y.right = n

    def transplant(self, u, v):
        if u.parent == None:
            self.root = v
        elif u == u.parent.left:
            u.parent.left = v
        else:
            u.parent.right = v

        if v != None:
            v.parent = u.parent

    def delete(self, z):
        if z.left == None:
            self.transplant(z, z.right)

        elif z.right == None:
            self.transplant(z, z.left)

        else:
            y = self.minimum(z.right)  # minimum element in right subtree
            if y.parent != z:
                self.transplant(y, y.right)
                y.right = z.right
                y.right.parent = y

            self.transplant(z, y)
            y.left = z.left
            y.left.parent = y

    def inorder(self, n):
        if n != None:
            self.inorder(n.left)
            print(n.data)
            self.inorder(n.right)

    # def treesort(arr):
    #     # Build BST
    #     if len(arr) == 0:
    #         return arr
    #     root = node(arr[0])
    #     for i in range(1, len(arr)):
    #         root.insert(arr[i])
    #     # Traverse BST in order.
    #     res = []
    #     inorder(root, res)
    #     return res


if __name__ == '__main__':
    t = BinarySearchTree()

    a = Node(10)
    b = Node(20)
    c = Node(30)
    d = Node(100)
    e = Node(90)
    f = Node(40)
    g = Node(50)
    h = Node(60)
    i = Node(70)
    j = Node(80)
    k = Node(150)
    l = Node(110)
    m = Node(120)

    t.insert(a)
    t.insert(b)
    t.insert(c)
    t.insert(d)
    t.insert(e)
    t.insert(f)
    t.insert(g)
    t.insert(h)
    t.insert(i)
    t.insert(j)
    t.insert(k)
    t.insert(l)
    t.insert(m)

    t.delete(a)
    t.delete(m)

    t.inorder(t.root)
    # t.search(t,150)